lib_nn/src/cpp/OutputTransformFn.cpp

```diff
@@ -0,0 +1,620 @@
+  stream.read(reinterpret_cast<char *>(&this->output_img_channels),
+              sizeof(int32_t));
+}
+
+void DirectWriteOutputTransform::Params::Serialize(std::ostream &stream) const {
+  stream.write(reinterpret_cast<const char *>(&this->output_img_channels),
+               sizeof(this->output_img_channels));
+}
+
+////////// DirectWriteOutputTransform //////////////
+DirectWriteOutputTransform::DirectWriteOutputTransform(const Params *params)
+    : params(params) {}
+
+int8_t *DirectWriteOutputTransform::output_transform_fn(
+    int8_t *Y, VPURingBuffer *acc, int32_t output_channel_group) {
+  const int32_t first_channel =
+      DirectWriteOutputTransform::ChannelsPerOutputGroup * output_channel_group;
+  const int32_t count =
+      std::min<int32_t>(DirectWriteOutputTransform::ChannelsPerOutputGroup,
+                        this->params->output_img_channels - first_channel);
+
+  // This might be slightly sped up with an xcore-specific implementation, but
+  // it's likely unnecessary
+  std::memcpy(Y, &acc->vR[0], count);
+
+  return &Y[count];
+}
+
+/******************************
+ * ShiftInt8OutputTransform
+ *****************************/
+
+////////// ShiftInt8OutputTransform::Params //////////////
+ShiftInt8OutputTransform::Params::Params(const int image_channels,
+                                         const int16_t shift)
+    : output_img_channels(image_channels) {
+  for (int k = 0; k < VPU_INT8_ACC_PERIOD; ++k) shifts[k] = shift;
+}
+
+ShiftInt8OutputTransform::Params::Params(const nn::ImageGeometry &output_image,
+                                         const int16_t shift)
+    : output_img_channels(output_image.depth) {
+  for (int k = 0; k < VPU_INT8_ACC_PERIOD; ++k) shifts[k] = shift;
+}
+
+ShiftInt8OutputTransform::Params::Params(std::istream &stream) {
+  int16_t shift;
+#define READ(X) stream.read(reinterpret_cast<char *>(&X), sizeof(X))
+  READ(this->output_img_channels);
+  READ(shift);
+#undef READ
+
+  for (int k = 0; k < VPU_INT8_ACC_PERIOD; ++k) shifts[k] = shift;
+}
+
+void ShiftInt8OutputTransform::Params::Serialize(std::ostream &stream) const {
+  auto shift = this->shifts[0];
+#define WRITE(X) stream.write(reinterpret_cast<const char *>(&X), sizeof(X));
+  WRITE(this->output_img_channels);
+  WRITE(shift);
+#undef WRITE
+}
+
+////////// ShiftInt8OutputTransform //////////////
+ShiftInt8OutputTransform::ShiftInt8OutputTransform(const Params *params)
+    : params(params) {}
+
+C_API void shift_int8_output_transform_ref(int8_t *output,
+                                           const VPURingBuffer *acc,
+                                           const int16_t *right_shifts,
+                                           const int channel_count) {
+  uint32_t write_mask = uint32_t((1LL << channel_count) - 1);
+
+  auto vpu = nn::VPU();
+  vpu_vector_t vec_tmp;
+  uint32_t tmp;
+
+  vpu.vldr(vpu_vect_0x80);
+  vpu.vstrpv(output, write_mask);
+
+  vpu.vldd(acc->vD);
+  vpu.vldr(acc->vR);
```

Comment:
Should these functions be similar to the ones above? Meaning VLDR() VLDD() used above?

Reply:
they are ok!